src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+
+use mentat_core::{
+    Attribute,
+    Entid,
+    HasSchema,
+    KnownEntid,
+    NamespacedKeyword,
+    TypedValue,
+    ValueType,
+};
+
+pub use mentat_core::attribute;
+use mentat_core::attribute::Unique;
+
+use ::{
+    CORE_SCHEMA_VERSION,
+    IntoResult,
+};
+
+use ::conn::{
+    InProgress,
+    Queryable,
+};
+
+use ::errors::{
+    ErrorKind,
+    Result,
+};
+
+use ::entity_builder::{
+    BuildTerms,
+    TermBuilder,
+    Terms,
+};
+
+pub use mentat_db::AttributeBuilder;
+
+pub type Version = u32;
+pub type Datom = (Entid, Entid, TypedValue);
+
+/// `Attribute` instances not only aren't named, but don't even have entids.
+/// We need two kinds of structure here: an abstract definition of a vocabulary in terms of names,
+/// and a concrete instance of a vocabulary in a particular store.
+/// Note that, because it's possible to 'flesh out' a vocabulary with attributes without bumping
+/// its version number, we need to know the attributes that the application cares about -- it's
+/// not enough to know the name and version. Indeed, we even care about the details of each attribute,
+/// because that's how we'll detect errors.
+#[derive(Debug)]
+pub struct Definition {
```

Comment:
`AbstractVocabulary` or `SymbolicVocabulary` vs. `ConcreteVocabulary`?

It rather feels like you're going to implement term rewriting like the transactor does, so parameterizing over `NamespacedKeywordOr<Entid>` might not be a bad approach.